File: packages/renderer-web/src/slider.ts
/**
 * Pure Slider math (ZAB-24) shared by the build pass, the arrange pass, the
 * input handlers and `setData`. No DOM — kept separate so it's unit testable
 * without a canvas, and so the Unity SDK has an unambiguous reference for the
 * same rules (the same role `toggle.ts` plays for the two-state control).
 *
 * The geometric model, in one place: the thumb's travel is the track minus the
 * thumb's own size, so the thumb never paints outside the node's rect; the fill
 * spans the whole fraction of the track, so it reaches the end at `max`.
 */

import type { SliderAxis } from "@zabloo/format";

/** Fallback arrow-key step when the slider is continuous: 5% of the range. */
const CONTINUOUS_STEPS = 20;

/** A slider's resolved range. `min`/`max` are author input, so they can be nonsense. */
export interface SliderRange {
  min: number;
  max: number;
  /** Quantization step, or 0 for a continuous value. */
  step: number;
}

/**
 * Resolves the declared range, defaulting to the unit interval. A degenerate
 * range (`max <= min`, NaN, a negative step) collapses to a fixed slider rather
 * than dividing by zero further down: it paints at its minimum and ignores
 * input, which is a visible authoring bug instead of a crash or a NaN rect.
 */
export function resolveRange(min: unknown, max: unknown, step: unknown): SliderRange {
  const lo = finite(min, 0);
  const hi = finite(max, 1);
  const declared = finite(step, 0);
  return {
    min: lo,
    max: Math.max(lo, hi),
    step: declared > 0 ? declared : 0,
  };
}

/**
 * Clamps into the range and snaps to `min + k * step` (no step = continuous).
 *
 * `max` is always a valid stop, even when the range is not a whole number of
 * steps (0..1 by 0.3 stops at 0.9): the player can SEE the end of the track, so
 * leaving it unreachable reads as a stuck control. The price is a short last
 * step, which is the smaller surprise.
 */
export function quantize(value: number, range: SliderRange): number {
  const clamped = Math.min(range.max, Math.max(range.min, finite(value, range.min)));
  if (range.step <= 0) return clamped;
  const steps = Math.round((clamped - range.min) / range.step);
  const snapped = Math.min(range.max, tidy(range.min + steps * range.step));
  return range.max - clamped <= clamped - snapped ? range.max : snapped;
}

/** Position of a value along the track, as a 0..1 fraction. */
export function fractionOf(value: number, range: SliderRange): number {
  const span = range.max - range.min;
  if (!(span > 0)) return 0;
  const clamped = Math.min(range.max, Math.max(range.min, finite(value, range.min)));
  return (clamped - range.min) / span;
}

/**
 * The value a pointer at `position` selects, given the track's own start and
 * length along its axis and the thumb's size. The travel is inset by half a
 * thumb at each end — the same inset `thumbStart` paints with — so the point
 * under the finger stays under the thumb's center through the whole drag
 * instead of drifting near the ends.
 *
 * On a vertical track the value grows upward (`min` at the bottom), like a
 * physical fader: `up` flips the axis mapping for that.
 */
export function valueAt(
  position: number,
  start: number,
  length: number,
  thumbSize: number,
  range: SliderRange,
  up = false,
): number {
  const travel = Math.max(0, length - thumbSize);
  const from = start + thumbSize / 2;
  const raw = travel > 0 ? (position - from) / travel : 0;
  const fraction = Math.min(1, Math.max(0, up ? 1 - raw : raw));
  return quantize(range.min + fraction * (range.max - range.min), range);
}

/**
 * The value one arrow-key press away, `direction` being +1 or -1 along the
 * slider's axis. A continuous slider borrows a step of 5% of its range, so the
 * keyboard is usable without forcing authors to declare a `step` they only
 * wanted for quantization.
 */
export function stepBy(value: number, direction: number, range: SliderRange): number {
  const span = range.max - range.min;
  if (!(span > 0)) return range.min;
  const step = range.step > 0 ? range.step : span / CONTINUOUS_STEPS;
  return quantize(quantize(value, range) + direction * step, range);
}

/** Distance from the track's start to the thumb's near edge, at that fraction. */
export function thumbStart(fraction: number, length: number, thumbSize: number): number {
  return Math.max(0, length - thumbSize) * fraction;
}

/**
 * How the fill and the thumb sit along the track, in track-relative px. The
 * fill spans the fraction of the FULL length (so `max` fills the rail to its
 * end) while the thumb moves inside the inset travel — the half-thumb gap at
 * the ends is what keeps every pixel inside the node's rect.
 *
 * On a vertical track both are measured from the bottom, so `start` is a
 * distance from the track's END; the caller flips it into view space.
 */
export interface SliderGeometry {
  fillLength: number;
  thumbStart: number;
}

export function sliderGeometry(
  fraction: number,
  length: number,
  thumbSize: number,
): SliderGeometry {
  const clamped = Math.min(1, Math.max(0, fraction));
  return {
    fillLength: length * clamped,
    thumbStart: thumbStart(clamped, length, Math.min(thumbSize, length)),
  };
}

/** True when this axis grows upward — the vertical fader's bottom-to-top mapping. */
export function growsUpward(axis: SliderAxis | undefined): boolean {
  return axis === "vertical";
}

function finite(value: unknown, fallback: number): number {
  return typeof value === "number" && Number.isFinite(value) ? value : fallback;
}

/**
 * Drops binary-float noise from a stepped value (`0.1 * 3` is
 * `0.30000000000000004`): the number travels to the game through the data
 * channel and may be shown as text, so a step of `0.1` must produce `0.3`. It
 * rounds far past any precision an author declares — it cleans up, it does not
 * quantize.
 */
function tidy(value: number): number {
  return Number.isInteger(value) ? value : Number(value.toFixed(10));
}

File: core/src/slider.cpp
#include <algorithm>
#include <cmath>

#include "slider.h"

namespace zabloo {
namespace {

/** Fallback arrow-key step when the slider is continuous: 5% of the range. */
constexpr double CONTINUOUS_STEPS = 20.0;

/**
 * Above this magnitude `tidy` leaves the number alone.
 *
 * Two reasons, and they meet: past ~1e5 a double's own spacing is wider than the
 * tenth decimal, so rounding there is already the identity map; and the `* 1e10`
 * the rounding is done with would leave the exactly-representable integers below
 * 2^53 and start rounding twice.
 */
constexpr double TIDY_LIMIT = 1e5;

/**
 * Drops binary-float noise from a stepped value (`0.1 * 3` is
 * `0.30000000000000004`): the number travels to the game through the data
 * channel and may be shown as text, so a step of `0.1` must produce `0.3`. It
 * rounds far past any precision an author declares — it cleans up, it does not
 * quantize.
 *
 * The reference spells this `Number(value.toFixed(10))`, which rounds halves
 * UP (toward +infinity) and not away from zero: hence `floor(v + 0.5)` and not
 * `std::round`, the same parity note the glyph snap carries (G4).
 */
double tidy(double value) {
  if (!std::isfinite(value)) return value;
  if (value == std::floor(value)) return value;  // `Number.isInteger`
  if (std::abs(value) >= TIDY_LIMIT) return value;
  return std::floor(value * 1e10 + 0.5) / 1e10;
}

double clamp_to(double value, const SliderRange &range) {
  return std::min(range.max, std::max(range.min, value));
}

}  // namespace

SliderRange resolve_range(const std::optional<double> &min, const std::optional<double> &max,
                          const std::optional<double> &step) {
  const double low = min.value_or(0.0);
  const double high = max.value_or(1.0);
  const double declared = step.value_or(0.0);
  SliderRange range;
  range.min = low;
  range.max = std::max(low, high);
  range.step = declared > 0.0 ? declared : 0.0;
  return range;
}

double quantize(double value, const SliderRange &range) {
  const double clamped = clamp_to(std::isfinite(value) ? value : range.min, range);
  if (range.step <= 0.0) return clamped;
  const double steps = std::floor((clamped - range.min) / range.step + 0.5);
  const double snapped = std::min(range.max, tidy(range.min + steps * range.step));
  return range.max - clamped <= clamped - snapped ? range.max : snapped;
}

double fraction_of(double value, const SliderRange &range) {
  const double span = range.max - range.min;
  if (!(span > 0.0)) return 0.0;
  const double clamped = clamp_to(std::isfinite(value) ? value : range.min, range);
  return (clamped - range.min) / span;
}

double value_at(double position, double start, double length, double thumb_size,
                const SliderRange &range, bool up) {
  const double travel = std::max(0.0, length - thumb_size);
  const double from = start + thumb_size / 2.0;
  const double raw = travel > 0.0 ? (position - from) / travel : 0.0;
  const double fraction = std::min(1.0, std::max(0.0, up ? 1.0 - raw : raw));
  return quantize(range.min + fraction * (range.max - range.min), range);
}

double step_by(double value, double direction, const SliderRange &range) {
  const double span = range.max - range.min;
  if (!(span > 0.0)) return range.min;
  const double step = range.step > 0.0 ? range.step : span / CONTINUOUS_STEPS;
  return quantize(quantize(value, range) + direction * step, range);
}

double thumb_start(double fraction, double length, double thumb_size) {
  return std::max(0.0, length - thumb_size) * fraction;
}

SliderGeometry slider_geometry(double fraction, double length, double thumb_size) {
  const double clamped = std::min(1.0, std::max(0.0, fraction));
  SliderGeometry out;
  out.fill_length = length * clamped;
  out.thumb_start = thumb_start(clamped, length, std::min(thumb_size, length));
  return out;
}

bool grows_upward(SliderAxis axis) { return axis == SliderAxis::Vertical; }

}  // namespace zabloo
